File: priorityQueue.js
class Node {
    constructor(val, priority) {
      this.val = val;
      this.priority = priority;
    }
  }



  class PriorityQueue {
    constructor() {
      this.values = [];
    }

    insert(node) {
      const v = this.values;
      v.push(node);

      let childIdx = v.length - 1;
      let parentIdx = ~~((childIdx - 1) / 2);

      while (v[parentIdx].priority > v[childIdx].priority) {
        [v[parentIdx], v[childIdx]] = [v[childIdx], v[parentIdx]];
        childIdx = parentIdx;
        parentIdx = ~~((childIdx - 1) / 2);
      }

    }

    // Recursion

    insertNode(node){
      const v = this.values;
      v.push(node);

      let childIdx = v.length - 1;
      let parentIdx = ~~((childIdx - 1) / 2);

      function checkPosition(arr, c, p) {
        if (arr[p].priority <= arr[c].priority) return;

        [arr[p], arr[c]] = [arr[c], arr[p]];

        let child = p;
        let par = ~~((child - 1) / 2);

        checkPosition(v, child, par);
      }

      checkPosition(v, childIdx, parentIdx);
    }


    removeHead() {
      const v = this.values;
      [v[0], v[v.length - 1]] = [v[v.length - 1], v[0]];
      const firstPriority = v.pop();

      let rootIdx = 0;

      let childIdx = this.getChild(v, rootIdx);

        while (!!v[childIdx] && v[rootIdx].priority > v[childIdx].priority) {
          [v[childIdx], v[rootIdx]] = [v[rootIdx], v[childIdx]];

          rootIdx = childIdx;

          childIdx = this.getChild(v, rootIdx);
        }

      return firstPriority;

    }


    // Recursive
    remove(){
      const v = this.values;
      [v[0], v[v.length - 1]] = [v[v.length - 1], v[0]];
      const firstPriority = v.pop();

      let rootIdx = 0;
      let childIdx = this.getChild(v, rootIdx);

      function checkPosition(arr, p, c, fn) {
        if (!arr[c] || (arr[p].priority <= arr[c].priority)) return;

        [arr[p], arr[c]] = [arr[c], arr[p]];

        let par = c;
        let ch = fn(v, par);

        checkPosition(v, par, ch, fn);
      }

      checkPosition(v, rootIdx, childIdx, this.getChild);

      return firstPriority;
    }


    getChild(arr, i){
      if (!!arr[(i * 2) + 1]) {
       return !arr[(i * 2) + 2] || (arr[(i * 2) + 1].priority < arr[(i * 2) + 2].priority) ?
        (i * 2) + 1 : (i * 2) + 2;
      }
      return null;
    }
  }



  let q = new PriorityQueue();
  q.values = [
    {
      val: "top",
      priority: 1
    },
    {
      val: "second",
      priority: 2
    },
    {
      val: "next",
      priority: 5
    },
    {
      val: "low",
      priority: 10
    }
  ];

  q.insertNode(new Node("new", 4));

  q.insertNode(new Node("newest", 3));

  q.remove();
  q.remove();

  console.log(q);

File: priority_queue.py
class Node:
    def __init__(self, val, priority):
        self.val = val
        self.priority = priority

class PriorityQueue:
    def __init__(self):
        self.values = []

    def insert(self, node):
        v = self.values
        v.append(node)

        child = len(v) - 1
        parent = (child - 1) // 2

        def check_pos(lst, p, c):
            if p < 0 or (lst[p].priority <= lst[c].priority): return

            lst[p], lst[c] = lst[c], lst[p]
            child = p
            parent = (child - 1) // 2

            check_pos(v, parent, child)

        check_pos(v, parent, child)

    def remove(self):
        v = self.values
        end = len(v) - 1

        v[0], v[end] = v[end], v[0]
        top_priority = v.pop()

        def get_child(lst, i):
            if len(lst) - 1 < (i * 2) + 1: return
            if len(lst) - 1 < (i * 2) + 2: return (i * 2) + 1

            return (i * 2) + 1 if lst[(i * 2) + 1].priority < lst[(i * 2) + 2].priority else (i * 2) + 2


        parent = 0
        child = get_child(v, parent)

        def check_pos(lst, p, c):
            if not c or lst[p].priority <= lst[c].priority: return

            lst[p], lst[c] = lst[c], lst[p]
            parent = c
            child = get_child(lst, parent)

            check_pos(v, parent, child)

        check_pos(v, parent, child)

        return top_priority


q = PriorityQueue()
q.insert(Node("first", 1))
q.insert(Node("next", 3))
q.insert(Node("then", 6))
q.insert(Node("low", 10))
q.insert(Node("second", 2))
q.insert(Node("mid", 8))
q.insert(Node("fourth", 4))

